### ISSUE TITLE
Add configuration setting to control whether a stack trace should be captured for each span

### ISSUE THREAD

**Issue by SergeyKleyman** [elastic/apm-agent-dotnet#307]
We need to decide 
- Would it make sense to add an additional parameter to `StartSpan/CaptureSpan` APIs to allow user to control it on more fine grained level?
- Should it be a simple boolean configuration setting resulting in the same behavior for all spans (unless overridden by a new parameter to public API) or a more sophisticated setting - for example list of span types for which stack trace should be captured? Maybe some aspect of a span other than its type to decide whether to capture stack trace - for example [Java agent uses span duration](https://www.elastic.co/guide/en/apm/agent/java/current/config-stacktrace.html#config-span-frames-min-duration). In case of the duration I assume the stack trace is for the end of the span not the start.
- What should be the default value?

**Comment by SergeyKleyman** [comment]
We can have a configuration setting (let's say something like `span_stack_trace_capture`) with the following values:
 - `disabled` (agent should not automatically capture stack trace for any span). Stack trace can still be captured via public API.
- `all_spans_on_start` (agent should capture stack trace for each span's start)
- `long_spans_on_end` (agent should capture stack trace for long spans' end where long can be controlled by another setting - `span_stack_trace_min_duration`)

**Comment by SergeyKleyman** [comment]
Indirectly related to https://github.com/elastic/apm-agent-dotnet/issues/225

**Comment by gregkalapos** [comment]
Run a perf. test to measure the perf. impact of capturing stacktraces. 

Test case: Creating 100 transaction, each with 10 spans.

Capturing stacktrace:
```

|                      Method |     Mean |    Error |   StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| Simple100Transaction10Spans | 84.18 ms | 1.665 ms | 3.402 ms |   5666.6667 |   2166.6667 |           - |            34.49 MB |

// * Legends *
  Mean                : Arithmetic mean of all measurements
  Error               : Half of 99.9% confidence interval
  StdDev              : Standard deviation of all measurements
  Gen 0/1k Op         : GC Generation 0 collects per 1k Operations
  Gen 1/1k Op         : GC Generation 1 collects per 1k Operations
  Gen 2/1k Op         : GC Generation 2 collects per 1k Operations
  Allocated Memory/Op : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 ms                : 1 Millisecond (0.001 sec)
```

Not capturing stacktraces:

```
|                      Method |     Mean |    Error |   StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| Simple100Transaction10Spans | 24.24 ms | 1.117 ms | 2.864 ms |   1937.5000 |    562.5000 |           - |            11.63 MB |
```

**Comment by SergeyKleyman** [comment]
@felixbarny mentioned that Java agent resolves stack traces in deferred fashion (not on app thread thus avoid blocking the app) - maybe it is worth measuring how much it will help if we resolve stack traces on payload thread instead.

**Comment by gregkalapos** [comment]
> @felixbarny mentioned that Java agent resolves stack traces in deferred fashion (not on app thread thus avoid blocking the app) - maybe it is worth measuring how much it will help if we resolve stack traces on payload thread instead.

Definitely something we can think about, but I think it won't be trivial. 

Here is how we do it now:
```
StackTrace = StacktraceHelper.GenerateApmStackTrace(new StackTrace(true).GetFrames(), _logger, $"Span `{Name}'");
```

In `GenerateApmStackTrace` (which is our method) we basically just create our own data structure.

The main part of the work is done by `new StackTrace(true).GetFrames()` and I suspect that this API already resolves the stack frames. 

I did 1 test run with only executing `new StackTrace(true)` - by only doing this we'd have the option to defer the rest of the work, so this would be the prerequisite to implement the suggestion and we'd do the rest of the work async later.

So I changed 
```
StackTrace = StacktraceHelper.GenerateApmStackTrace(new StackTrace(true).GetFrames(), _logger, $"Span `{Name}'");
```

To 

```
new StackTrace(true);
```

And here is the result:

```

|                      Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| Simple100Transaction10Spans | 50.93 ms | 0.9664 ms | 0.8567 ms |   5300.0000 |   1000.0000 |           - |            31.82 MB |

```

As you can see, this already caused most of the allocation, and the runtime overhead also increased drastically. So I think if we want to implement the suggestion we need to look for another API, because this one seems to resolve everything immediately.

**Comment by gregkalapos** [comment]
Yeah, the API we use resolves everything synchronously

https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs#L58

https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/src/System/Diagnostics/StackFrame.CoreCLR.cs#L29

**Comment by SergeyKleyman** [comment]
What about `new StackTrace(false)` instead of `new StackTrace(true)` ?
According to https://ayende.com/blog/3879/reducing-the-cost-of-getting-a-stack-trace `new StackTrace(false)` might be quite faster

**Comment by SergeyKleyman** [comment]
> As you can see, this already caused most of the allocation

Of course, allocations number will not reduced by deferring stack trace resolution (even if `new StackTrace(true)` didn't make most of allocations anyway). The goal of deferring is to minimize immediate latency incurred by application.

**Comment by gregkalapos** [comment]
> What about new StackTrace(false) instead of new StackTrace(true) ?
According to https://ayende.com/blog/3879/reducing-the-cost-of-getting-a-stack-trace new StackTrace(false) might be quite faster

That's definitely faster, but then we won't have line numbers. The `true` was added by a community PR, so some people prefer to have that info.

> > As you can see, this already caused most of the allocation

> Of course, allocations number will not reduced by deferring stack trace resolution (even if new StackTrace(true) didn't make most of allocations anyway). The goal of deferring is to minimize immediate latency incurred by application.

I suspect that resolving offsets to strings would still allocate. But my point with this was that there is no way to reduce the current code in a way that we defer most of this work from the user thread to minimize latency - by executing `new StackTrace(true)` we already do most of the work, if we don't even do that, then... well then there is nothing left from capturing the callstack.

I imagine that they key here would be to just capture the offsets to the class/method names and do the resolution of those and the transformation to our own data structure on a different thread - could be e.g. on the thread of `PayloadSender`. But as it seems `StackTrace` does not offer this functionality - at least I don't see it.

**Comment by SergeyKleyman** [comment]
> That's definitely faster, but then we won't have line numbers.

Maybe it's worth a configuration option? This way user can choose 
1. No stack trace
2. Stack trace with method names and line numbers but with performance overhead
3. Stack trace with method names only but with much less performance overhead

Otherwise it's a choice between (1) and (2) and performance overhead in (2) might be prohibitive.

**Comment by gregkalapos** [comment]
My original plan was to implement the same settings that other agents have (probably should have started with this before my rant about overhead 🤣 ) 

- [`ELASTIC_APM_STACK_TRACE_LIMIT`](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-stack-trace-limit)
- [`ELASTIC_APM_SPAN_FRAMES_MIN_DURATION `](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-span-frames-min-duration-ms)

Asking other agents if they have similar setting to control line number collection. I personally think that the 2 settings above should be enough.

**Comment by gregkalapos** [comment]
Also to address this:

> Would it make sense to add an additional parameter to StartSpan/CaptureSpan APIs to allow user to control it on more fine grained level?

I think we are ok with a simple config here. If we'd do this then suddenly by let's say turning off stack traces globally, users would still have stack traces to some spans that are captured by the API and manually forced to capture stacktraces. In bigger organisations I think it's fair to assume that not everyone knows what the devs coded into the codebase - and offering this option would cause confusions. 

Do we have other agents doing this? If not, I think we should not do it either.

**Comment by SergeyKleyman** [comment]
If we assume that `ELASTIC_APM_STACK_TRACE_LIMIT` purpose is performance overhead control it might not be that useful from what you measured. So having an alternative configuration option that is useful to control performance overhead of .NET agent current implementation is IMHO a valid reason to break alignment with other agents.

**Comment by gregkalapos** [comment]
> If we assume that `ELASTIC_APM_STACK_TRACE_LIMIT` purpose is performance overhead control it might not be that useful from what you measured. So having an alternative configuration option that is useful to control performance overhead of .NET agent current implementation is IMHO a valid reason to break alignment with other agents.

Why so? 

It says: 
> Setting the limit to 0 will disable stack trace collection, while any positive integer value will be used as the maximum number of frames to collect.

I thought if users can't accept the overhead they set it to 0 and they will have the performance of my 2. measurement from [this comment](https://github.com/elastic/apm-agent-dotnet/issues/307#issuecomment-510134603).

**Comment by SergeyKleyman** [comment]
If the only useful use case for `ELASTIC_APM_STACK_TRACE_LIMIT` from .NET agent prospective is to switch stack trace collection on/off then the same can be achieved [with `ELASTIC_APM_SPAN_FRAMES_MIN_DURATION` alone](https://www.elastic.co/guide/en/apm/agent/java/current/config-stacktrace.html#config-span-frames-min-duration):
> To disable stack trace collection for spans completely, set the value to 0ms.

Obviously `ELASTIC_APM_STACK_TRACE_LIMIT` is more than just switching stack trace collection on/off. It's might be to control performance overhead on the agent side and/or network traffic/storage on the backend side. If it's mostly the former and if it won't achieve the desired goal for .NET agent then we can consider an alternative configuration option in addition to or instead of `ELASTIC_APM_STACK_TRACE_LIMIT`.

**Comment by gregkalapos** [comment]
From what I see in the CoreCLR source code linked in [this comment](https://github.com/elastic/apm-agent-dotnet/issues/307#issuecomment-510147218) the number of frames you capture make a difference in terms of performance. Look [here](https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/src/System/Diagnostics/StackFrame.CoreCLR.cs#L39).

So from what I understand, the use case of `ELASTIC_APM_STACK_TRACE_LIMIT` is indeed limiting the number of frames in the stack trace, so it's more than just on/off... setting it to e.g. `10` is different than letting it be unlimited - and that seems to be true from the perspective of the overhead when we capture the stack trace.

Happy to measure later, but I think we are ok here to align with other agents.

**Comment by SergeyKleyman** [comment]
I'm not sure there is a way to implement `ELASTIC_APM_STACK_TRACE_LIMIT` as performance controlling option using .NET's `StackTrace` class. `StackTrace` class has `StackTrace (int skipFrames)` constructor but `skipFrames` is the opposite of `ELASTIC_APM_STACK_TRACE_LIMIT` - `skipFrames` excludes N frames from the top of the stack (assuming stack grows upward) while `ELASTIC_APM_STACK_TRACE_LIMIT` takes up to N frames from the top.

**Comment by gregkalapos** [comment]
> I'm not sure there is a way to implement `ELASTIC_APM_STACK_TRACE_LIMIT` as performance controlling option using .NET's `StackTrace` class. `StackTrace` class has `StackTrace (int skipFrames)` constructor but `skipFrames` is the opposite of `ELASTIC_APM_STACK_TRACE_LIMIT` - `skipFrames` excludes N frames from the top of the stack (assuming stack grows upward) while `ELASTIC_APM_STACK_TRACE_LIMIT` takes up to N frames from the top.

Och, that's true and very unfortunate. Ok, then the argument of controlling the overhead in that terms may not hold.

I still think we should first try to implement what we have in the [ spreadsheet](https://docs.google.com/spreadsheets/d/1JJjZotapacA3FkHc2sv_0wiChILi3uKnkwLTjtBmxwU/edit#gid=0)

**Comment by SergeyKleyman** [comment]
We can implement `ELASTIC_APM_STACK_TRACE_LIMIT` and `ELASTIC_APM_SPAN_FRAMES_MIN_DURATION` but if the initial point of this issue was to control performance overhead and since those options don't do that then I would recommend implementing additional configuration option(s) that do control performance overhead for .NET agent.

**Comment by gregkalapos** [comment]
Ok, I'll implement those two. 

I think those still will have the desired effect - you can turn the whole thing off and that's fine in the 1. round. We can discuss having an additional setting in a broader meeting, eg. in the next planning.

**Comment by SergeyKleyman** [comment]
@gregkalapos What about #225 - it seems that this issue supersedes it?

**Comment by gregkalapos** [comment]
Right, closed it.

**Comment by gregkalapos** [comment]
Implemented in #374.
